### ISSUE TITLE
Limit `CompoundAccessControlCondition` to prevent overloading Ursulas

### ISSUE THREAD

**Issue by derekpierre** [nucypher/nucypher#3148]
Related to discussion in #3140 raised by @piotr-roslaniec - https://github.com/nucypher/nucypher/pull/3140#discussion_r1230923557.

What sort of limits should we put on `CompoundAccessControlConditions`.

```
@[piotr-roslaniec](https://github.com/piotr-roslaniec) piotr-roslaniec [21 minutes ago]
Not specific to this PR:

Should we put a ceiling on the number of elements in conditions? Something in the order of ~100. We could short-circuit it too: evaluate the first 100, and if there exists another then return False.
Seems like a possible DoS vector. WDYT?

@[KPrasch](https://github.com/KPrasch) KPrasch [16 minutes ago]
That's a very sensible restriction, in fact I'd make the ceiling much lower, perhaps 5-10 elements since each call adds o(n) runtime at minimum, and maps directly onto node operational costs. Even beyond that, we may consider some form of rate limiting. All of which needs to be addressed in another PR IMO.

@[derekpierre](https://github.com/derekpierre) derekpierre [13 minutes ago]
Nice thought!

We can potentially limit the length of the list of conditions, and raise an InvalidCondition exception if too long during object creation/schema validation.

I'd rather do that than allow the object to be created then fail during execution because it is too long - could be too confusing to the user. That being said, technically you can work around it by having repeated nested compound conditions each with operands under the limit.

So maybe some kind of overall count needs to be calculated during creation...?

Some good thoughts here. Let's tackle it in another PR.


@[piotr-roslaniec](https://github.com/piotr-roslaniec) piotr-roslaniec [6 minutes ago]
There's definitely both client and server validation involved there, I think this is often the case with the condition module as a whole.

I agree that's material for another contribution. I'll be interested in scoping this, also for the purpose of implementation in nucypher-ts.
```

**Comment by jMyles** [comment]
This is really two things, right?

One is providing a compound condition such that evaluation requires hammering of the provider.  The other is providing a compound condition with strenuous CPU (or even RAM) overhead.

Is that correct?

If so, it seems that we may want two calming strategies rather than one.

As to @derekpierre's question about whether this is 7-worthy:

Obviously a situation in which we have so much _genuine_ engagement in the network that our nodes are overwhelmed will be a wonderful problem to have. An insincere happening like DoS, not so much. But the mechanics of this optimization are touchy.

So I think that if we are going to do something for version 7.0.0, then:

* Let's do something really naive and simple, like @KPrasch's bytes suggestion
* Let's communicate that additional optimizations - which will take the form of restrictions - are coming.

**Comment by derekpierre** [comment]
> One is providing a compound condition such that evaluation requires hammering of the provider. The other is providing a compound condition with strenuous CPU (or even RAM) overhead.

> Is that correct?

Our conditions are all on-chain so this is the number of provider requests concern
Nesting of compound conditions (compound conditions within compound conditions) is more the CPU concern

That being said, having some limit reasonably mitigates both issues.

> Let's do something really naive and simple, like @KPrasch's bytes suggestion

That is the easiest solution thus far. It's just a question of determining what overall bytes size cap for a compound condition should be. 

Probably just create some "reasonably expected" test compound condition with `x` total conditions (probably pick the largest condition type, contract maybe(?)) and includes `y` logical nestings and then use that overall byte size as the cap.

Determining `x` and `y` for the test condition to use to determine the cap is the bulk of the work.

**Comment by derekpierre** [comment]
Implemented limit on number of contained conditions and limit on nesting of multi-condition types via https://github.com/nucypher/nucypher/pull/3500

```
- Limit on number of conditions for a multi-condition type. For now the list size is capped at length 5
- Limit nesting of multi-condition types including combinations of nesting of each other. The limit is set to 2 levels of nesting i.e. you can have a multi condition type that includes a sub multi-condition type but that sub multi-condition cannot then include another multi-condition type.
- 
Basically in the worst case you have a multi-condition, that has all 5 conditions in the list be multi-conditions, meaning that max number of conditions to evaluate for any multi-condition (based on current limits) will be 5 * 5 = 25 conditions
```
